[RFC] Add ‘eob’ option to fillchars

FriedSock opened the pull request:
This option allows configuring what character is shown on the empty
lines at the end of a buffer, previously hardcoded to ‘~’

Fixes: https://github.com/neovim/neovim/issues/4181

justinmk commented on test/functional/options/fillchars_spec.lua:
```diff
@@ -0,0 +1,39 @@
+local Screen = require('test.functional.ui.screen')
+local clear, execute = helpers.clear, helpers.execute
+
+if helpers.pending_win32(pending) then return end
```
This should work just fine on Windows.

justinmk commented on test/functional/options/fillchars_spec.lua:
```diff
@@ -0,0 +1,39 @@
+    screen:detach()
+  end)
+
+  describe('the "eob" option', function()
```
describe('"eob" flag', function()

FriedSock commented:
I noticed a mistake in the documentation/implementation whilst doing this PR. The docs for `'fillchars'` state:
```
	Characters to fill the statuslines and vertical separators.
	It is a comma separated list of items:

	  item		default		Used for ~
	  stl:c		' ' or '^'	statusline of the current window
	  stlnc:c	' ' or '='	statusline of the non-current windows
	  vert:c	'|'		vertical separators |:vsplit|
	  fold:c	'-'		filling 'foldtext'
	  diff:c	'-'		deleted lines of the 'diff' option
	  eob:c		'~'		empty lines at the end of a buffer

	*Any one that is omitted will fall back to the default.  For "stl" and
	"stlnc" the space will be used when there is highlighting, '^' or '='
	otherwise.*
```

However, if one of the items is missing, it will not fall back to the default. This behaviour also exists in vim 7.

justinmk commented on test/functional/options/fillchars_spec.lua:
```diff
@@ -0,0 +1,39 @@
+        ~                        |
+                                 |
+      ]])
+      execute('set fillchars+=eob:\\ ')
```
how about trying some other characters? Including multibyte characters, if those are supported.

justinmk replied:
what happens if it's set to empty string?

FriedSock replied:
multibyte chars aren't supported, neither is setting anything in fillchars to empty string.

bfredl replied:
multibyte chars seem to work fine (maybe you meant composing chars won't)

FriedSock replied:
Ah yes sorry, I mistakenly tried this with multiple characters. Have added a multibyte character test case now.

justinmk commented:
LGTM after comments. Seems easy enough.

jamessan commented:
Thanks for working on this @FriedSock.  I would highly recommend submitting this patch upstream, as well.

FriedSock committed:
Add ‘eob’ option to fillchars

This option allows configuring what character is shown on the empty
lines at the end of a buffer, previously hardcoded to ‘~’

FriedSock committed:
Fix linting errors

The linter picked up on adding  `{&fill_eob, “eob”}` to the filltab
array; make the rest of the lines in the filltab and lcstab struct
arrays consistent as well.

justinmk commented:
- I don't think it works with multibyte characters, actually. E.g., try this: `Ｘ`
    - `:help 'listchars'` but `:help 'fillchars'` suggests only "stl" and "stlnc" flags must be single-byte. 
    - Probably `win_draw_end()` needs to be fixed.
- Need a note in `vim_diff.txt`, see the 'cpoptions' line for example.

ZyX-I commented on test/functional/options/fillchars_spec.lua:
```diff
@@ -0,0 +1,45 @@
+        ñ                        |
+        ñ                        |
+        :set fillchars+=eob:ñ    |
+      ]])
```
I would suggest also checking fullwidth (e.g. U+FF21), composing (both `a`+U+0340 and just U+0340), invalid unicode (`let &fillchars="eob:\x80"`), control characters (`let &fillchars="eob:\n"`), Unicode control characters (`let &fillchars="eob:\u0094"`), just two characters (`ab`). Note: I am not suggesting that all of this must work. Checking errors is also needed. Especially interesting just U+0340: try using `let &fillchars="diff:\u0340"` and running a diff, result is weird.

ZyX-I replied:
And missing test for empty fillchars, especially empty after successfully setting to something non-standard.

LeOtaku commented:
Any reason this has been delayed for so long? What is preventing this from being pushed forward?

KillTheMule commented:
The author hasn't been active for some time. Someone needs to pick this up and finish it.

bfredl commented:
I might be looking into this. I will also consider converting fillchars to schar_T (#7992) representation which maybe makes the code simpler and will allow composing chars as fillchars (mostly because why not), though if this is too messy in a separate PR.

bfredl commented:
Rebased as #8546

justinmk commented:
Closed by https://github.com/neovim/neovim/pull/8546
